modules/oe_content_persistent/src/ContentUuidResolver.php

```diff
@@ -0,0 +1,224 @@
+    $this->cache = $cache;
+    $this->supportedStorages = $supported_storages;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setCacheKey($key) {
+    $this->cacheKey = $this->getCachePrefixKey() . $key;
+  }
+
+  /**
+   * Getting current cache key.
+   */
+  public function getCacheKey() {
+    return $this->cacheKey;
+  }
+
+  /**
+   * Clear static cache.
+   */
+  public function resetStaticCache(): void {
+    $this->uuidCacheTags = [];
+    $this->lookupMap = [];
+    $this->cacheNeedsWriting = FALSE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function writeCache() {
+    // We assume that better to write cache only on request finishing
+    // on Controller page.
+    if ($this->cacheNeedsWriting === TRUE && !empty($this->cacheKey) && !empty($this->lookupMap)) {
+      $twenty_four_hours = 60 * 60 * 24;
+
+      if ($cache_tags = reset($this->uuidCacheTags)) {
+        $this->cache->set($this->cacheKey, reset($this->lookupMap), $this->getRequestTime() + $twenty_four_hours, $cache_tags);
+      }
+      else {
+        $this->cache->set($this->cacheKey, reset($this->lookupMap), $this->getRequestTime() + $twenty_four_hours);
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAliasByUuid(string $uuid, string $langcode = NULL): ?string {
+    $langcode = $langcode ?: $this->languageManager->getCurrentLanguage(LanguageInterface::TYPE_URL)->getId();
+
+    // Here we are trying to use static cache.
+    if (!empty($this->lookupMap[$uuid])) {
+      return $this->lookupMap[$uuid];
+    }
+
+    // Using cache key initialized in controller
+    // (if we use service inside controller).
+    if ($this->cacheKey) {
+      if ($cached = $this->cache->get($this->cacheKey)) {
+        // Used for bubble up cache tags to page cache level.
+        $this->addCacheTags($cached->tags);
+        return $cached->data;
+      }
+      else {
+        // Informing about cache writing on kernel termination.
+        // @see \Drupal\oe_content_persistent\EventSubscriber\UuidPathSubscriber and $this->writeCache().
+        $this->cacheNeedsWriting = TRUE;
+      }
+    }
+    // Reuse previously cached alias of uuid,
+    // even if we don't use this service with controller.
+    elseif ($cached = $this->cache->get($this->getCachePrefixKey() . $uuid)) {
+      return $cached->data;
+    }
+
+    // Try to retrieve entities from supported storages.
+    foreach ($this->supportedStorages as $storage_type) {
+      $storage = $this->entityTypeManager->getStorage($storage_type);
+      // Retrieving correct entity by uuid as we can't get entity internal url
+      // without loading full entity.
+      $entities = $storage->loadByProperties(['uuid' => $uuid]);
+      if (empty($entities)) {
+        return NULL;
+      }
+
+      $entity = reset($entities);
+      if ($entity instanceof EntityInterface) {
+        $alias = $this->aliasManager->getAliasByPath('/' . $entity->toUrl()->getInternalPath(), $langcode);
```

Comment:
I would consider providing a link template so you can directly get the uuid path from an entity. e.g. $entity->url('purl')

Reply:
Thanks for remark. I will do this in related ticket regarding linking in textarea fields.